### DISCUSSION TITLE
CI agent on Linux ARM64

### DISCUSSION THREAD

**Discussion by martin-g** [su2code/SU2#1611] [General]
Hello,

I am new to the community and I am interested to run SU2 on Linux ARM64.

I've just noticed https://github.com/su2code/SU2/issues/1609 (issue) and https://github.com/su2code/SU2/issues/1610 (PR with a fix).

I see the project uses Github Actions as CI (https://github.com/su2code/SU2/pull/1610/checks).
Unfortunately Github Actions has only x86_64 build nodes in the cloud. Linux ARM64 is supported but only for self-hosted build nodes but those are not recommended for public repositories like this one.

What are the project's options to add CI on ARM64 ?
My employer could provide a VM for long term usage if there is an interest!

Regards,
Martin

**Comment by pcarruscag** [comment]
Hello, we can talk about this in more detail in today's developer meeting https://meet.jit.si/SU2_DevMeeting at 3pm UK (if you can join).

**Comment by martin-g** [comment]
Thanks for the invitation!
I will try to join !

**Comment by martin-g** [comment]
Hi @pcarruscag !

I thought a bit how to setup the ARM64 builder node and came with the following idea:

1) Setup a self-hosted runner
2) Duplicate https://github.com/su2code/SU2/blob/master/.github/workflows/regression.yml
3) Change the [on](https://github.com/su2code/SU2/blob/42d157143f6d11fc7393c17fd4b6bb8ade30ce6b/.github/workflows/regression.yml#L3) part to be a `scheduled` run only on `master` branch.
4) Replace `ubuntu-latest` with the self-hosted runner in all `runs-on`

I.e. the ARM64 self-hosted runner will be used for post-validation only. It will execute only approved and merged code to `master`.
There are not that many commits per day, so the post-validation should be still useful. If needed it may be scheduled several times per day to shrink the time window and make it easier to pin point which merge caused a regression.

The only drawback I see is that we will have to duplicate the code in regression.yml. I see no way how to reuse it.

WDYT ?

**Reply by pcarruscag** [reply]
Hi Martin,
Thank you for looking into that, sounds like a good strategy.
The only detail is that we should target develop instead of master, because we cut new releases from develop and then update master.
Since we would be copying the script we could run just on merges into develop, and not on pull requests to reduce overhead.

We can also look into cross compiling for ARM as part of the release management workflow, e.g. https://github.com/su2code/SU2/runs/6111756542?check_suite_focus=true

**Reply by martin-g** [reply]
Hi Pedro!

I will work on this and send a PR soon!


> Since we would be copying the script we could run just on merges into develop, and not on pull requests to reduce overhead.

What do you mean here ? 
I have the feeling you suggest to remove this: https://github.com/su2code/SU2/blob/42d157143f6d11fc7393c17fd4b6bb8ade30ce6b/.github/workflows/regression.yml#L8-L11
I think the PR validation (on x86_64) would be still useful!

**Reply by pcarruscag** [reply]
Yes, but if you end up with a completely different script for ARM, then we may not need to run on every update of a pull request.
Maybe I got it wrong and you only want to duplicate part of the script inside itself?

**Reply by martin-g** [reply]
Let me try and see what kind of changes would be needed!

**Reply by martin-g** [reply]
How did you build https://hub.docker.com/r/su2code/build-su2 ?
At the moment it is linux/amd64 only.
Did you push it manually or it is automated somewhere ?

**Reply by pcarruscag** [reply]
Hi Martin, the docker scripts are here https://github.com/su2code/Docker-Builds
